pkg/apis/projectcalico/v3/nodestatus.go

```diff
@@ -0,0 +1,201 @@
+	Status CalicoNodeStatusStatus `json:"status,omitempty" protobuf:"bytes,2,opt,name=status"`
+}
+
+// CalicoNodeStatusSpec contains the specification for a CalicoNodeStatus resource.
+type CalicoNodeStatusSpec struct {
+	// The node name identifies the Calico node instance for node status.
+	Node string `json:"node,omitempty" validate:"required,name"`
+
+	// Classes specifies type of information CalicoNodeStatus should contain.
+	Classes []NodeStatusClassType `json:"classes,omitempty"`
+
+	// UpdateIntervalInSeconds is the period at which CalicoNodeStatus should be updated.
+	// Set to 0 to disable CalicoNodeStatus refresh. [Default: 10]
+	UpdateIntervalInSeconds int `json:"updateIntervalInSeconds,omitempty"`
+}
+
+// CalicoNodeStatusStatus defines the observed state of CalicoNodeStatus.
+// No validation needed for status since it is updated by Calico.
+type CalicoNodeStatusStatus struct {
+	// LastUpdated is a timestamp representing the server time when CalicoNodeStatus object
+	// last updated. It is represented in RFC3339 form and is in UTC.
+	// +nullable
+	LastUpdated metav1.Time `json:"lastUpdated,omitempty"`
+
+	// AdditionalInfo is a a human-readable description of the status of last update.
+	AdditionalInfo string `json:"additionalInfo,omitempty"`
+
+	// Agent holds agent status on the node.
+	Agent CalicoNodeAgentStatus `json:"agent,omitempty"`
+
+	// BGP holds node BGP status.
+	BGP CalicoNodeBGPStatus `json:"bgp,omitempty"`
+
+	// Route represents routes on the node.
+	Route CalicoNodeRouteStatus `json:"route,omitempty"`
+}
+
+// CalicoNodeAgentStatus defines the observed state of agent status on the node.
+type CalicoNodeAgentStatus struct {
+	// Bird4 represents the latest observed status of bird4.
+	Bird4 CalicoNodeBirdStatus `json:"bird4,omitempty"`
+
+	// Bird6 represents the latest observed status of bird6.
+	Bird6 CalicoNodeBirdStatus `json:"bird6,omitempty"`
+}
+
+// CalicoNodeBGPStatus defines the observed state of BGP status on the node.
+type CalicoNodeBGPStatus struct {
+	// The total number of established bgp sessions.
+	NumEstablished int `json:"numEstablished,omitempty"`
+
+	// The total number of non-established bgp sessions.
+	NumNotEstablished int `json:"numNotEstablished,omitempty"`
+
+	// V4Peers represents IPv4 BGP peers status on the node.
+	V4Peers []CalicoNodePeer `json:"v4Peers,omitempty"`
+
+	// V6Peers represents IPv6 BGP peers status on the node.
+	V6Peers []CalicoNodePeer `json:"v6Peers,omitempty"`
+}
+
+// CalicoNodeRouteStatus defines the observed state of routes status on the node.
+type CalicoNodeRouteStatus struct {
+	// V4 represents IPv4 routes on the node.
+	V4 []CalicoNodeRoute `json:"bird4,omitempty"`
+
+	// V6 represents IPv6 routes on the node.
+	V6 []CalicoNodeRoute `json:"bird6,omitempty"`
+}
+
+// CalicoNodeBirdStatus defines the observed state of bird.
+type CalicoNodeBirdStatus struct {
+	// Ready indicates if bird status is ready.
+	Ready bool `json:"ready,omitempty"`
```

Comment:
API conventions say to avoid using boolean fields for things like status.

Instead, we should use an enum.

e.g.,

```yaml
status: Ready | NotReady
```

Reply:
This comment still needs to be addressed.